### PR TITLE
Collection page restriction & validations

### PR DESCRIPTION
**Pull request by hale4029**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

- Update request spec - stub user session vs. using params
- Add validates uniqueness of a user's collection title
- Collection index page logic for logged out user 

## Related Tickets & Documents

#13 

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

### CODE REVIEW

File: spec/models/user_collection_spec.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@
 RSpec.describe UserCollection, type: :model do
   describe "validations" do
     it { is_expected.to validate_presence_of :title }
+    it { should validate_uniqueness_of(:title).scoped_to(:user_id).case_insensitive }
```
Comment (tyladevon) [inline spec/models/user_collection_spec.rb:6]: I kept this validation over the one I wrote as it is more thorough

File: app/controllers/collections_controller.rb [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
   skip_forgery_protection
-  
+
   def index
-    @collections = UserCollection.where(user_id: session_current_user_id)
+    if session_current_user_id
+      @collections = UserCollection.where(user_id: session_current_user_id)
+    else
+      @collections = false
+    end
```
Comment (ZaneMeroff) [inline app/controllers/collections_controller.rb:9]: Happy we have this in place so we can check if there is a user logged in.

File: app/models/user_collection.rb [hunk 3]
```diff
@@ -3,4 +3,6 @@ class UserCollection < ApplicationRecord
   belongs_to :user
   has_many :article_collections
   has_many :articles, through: :article_collections
+  validates :title, presence: true
+  validates_uniqueness_of :title, scope: :user_id, case_sensitive: false
```
Comment (ZaneMeroff) [inline app/models/user_collection.rb:7]: Great work implementing a catch so a user can't add duplicate articles in the same collection!

File: app/views/collections/index.html.erb [hunk 4]
```diff
@@ -55,3 +56,6 @@
     </a>
   <% end %>
 </div>
+<% else  %>
+<h1>Oops, You Must Login To See Your Collections</h1>
+<% end %>
```
Comment (ZaneMeroff) [inline app/views/collections/index.html.erb:61]: Glad we have this! Maybe we should reword it to also address those who do not have a username?

File: db/migrate/20200323191741_create_user_collections.rb [hunk 5]
```diff
@@ -3,7 +3,9 @@ def change
     create_table :user_collections do |t|
       t.string :title
       t.references :user, foreign_key: true
-      t.timestamps 
+      t.timestamps
+
+      add_index(:user_collections, :user)
```
Comment (ZaneMeroff) [inline db/migrate/20200323191741_create_user_collections.rb:8]: Glad we have timestamps here. I think it's always a good idea have this in place to keep track.
